### PR TITLE
Added support for new config option for taking device private key location

### PR DESCRIPTION
**Pull request by HarshGandhi-AWS**
*Issue #101*

*Description of changes:*
With this commit, I am adding config option for user to provide the device private key location via CLI or JSON file. This option will be required for provisioning the device using CSR file (CreateCertificateFromCsr API). If the device key or the CSR file is not provided, Device Client will use claim certificate and bootstrap private key (CreateKeysAndCertificate API) for provisioning the device.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -558,9 +558,9 @@ More details about AWS IoT Fleet Provisioning by claim can be found here: https:
 
 *Note: If the fleet provisioning feature fails to provision the new key, certificate or thing/device, the device client will abort with fatal error.*
 
-*Note: If a CSR file is not provided, the Device Client will use Claim Certificate and Private key for provisioning the device.*
+*Note: If the CSR file or the Device Private Key is not provided, the Device Client will use Claim certificate and Private key for provisioning the device*
```
Comment (matthewpaul) [inline README.md:561]: Shouldn't this say - "if the CSR file is specified without also specifying a private key, the Device Client will..."

File: README.md [hunk 2]
```diff
@@ -558,9 +558,9 @@ More details about AWS IoT Fleet Provisioning by claim can be found here: https:
...
 
-*Note: Please make sure that the Claim certificate, private key and/or CSR file stored on device side have respective permissions applied to it as mentioned above in the "File and Directory Permissions" section of the readme.*
+*Note: Please make sure that the Claim certificate, private key and/or CSR file, device private key stored on device side have respective permissions applied to it as mentioned above in the "File and Directory Permissions" section of the readme.*
```
Comment (matthewpaul) [inline README.md:563]: the and/or should be moved to right before "device private key"

**Review by matthewpaul — COMMENTED** [review]
Logic looks good to me, might just need a few updates on the documentation language. Also, do we need to update:

1. The setup script
2. The configuration template file
3. Test cases

File: source/fleetprovisioning/FleetProvisioning.cpp [hunk 3]
```diff
@@ -551,6 +555,34 @@ bool FleetProvisioning::ProvisionDevice(shared_ptr<SharedCrtResourceManager> fpC
...
+{
+    struct stat info;
+    bool locatedDeviceKey = true;
+    string expandedPath = FileUtils::ExtractExpandedPath(filePath.c_str());
+    if (stat(expandedPath.c_str(), &info) != 0)
+    {
+        LOGM_ERROR(
+            TAG,
+            "*** %s: Failed to find the device private key %s, file does not exist ***",
+            DeviceClient::DC_FATAL_ERROR,
```
Comment (matthewpaul) [inline source/fleetprovisioning/FleetProvisioning.cpp:568]: Is this actually a fatal error? I thought that if we can't find the private key, we are supposed to default back to the other FP API?

Comment (HarshGandhi-AWS) [reply source/fleetprovisioning/FleetProvisioning.cpp:568]: My bad. I'll update the log.

File: test/config/TestConfig.cpp [hunk 4]
```diff
@@ -33,6 +33,12 @@ TEST(Config, AllFeaturesEnabled)
     "device-defender": {
         "enabled": true,
         "interval": 300
+    },
+    "fleet-provisioning": {
+        "enabled": true,
+        "template-name": "template-name",
+		"csr-file": "csr-file",
```
Comment (matthewpaul) [inline test/config/TestConfig.cpp:40]: It looks like the formatting here is a bit weird, csr-file and device-key should probably be aligned with enabled and template-name

**Review by matthewpaul — APPROVED** [review]

File: test/config/TestConfig.cpp (hunk 4 above)
Comment (HarshGandhi-AWS) [reply test/config/TestConfig.cpp:40]: Looks like an tabs vs space issue (looks perfect in text editor). The change is required in every test case and will have a separate PR to resolve the issue.

**Review by lwilkovich — APPROVED** [review]

**Pull request merged**